File: PYTHON/Data Structures/3. sets.py
# https://www.w3schools.com/python/python_sets_methods.asp

# O(1)
# add() - Adds an element to the set

# O(1)
# in

# O(1)
# len()

# O(N)
# copy() - Returns a copy of the set
# O(N)
# clear() - Removes all the elements from the set

# O(1)
# discard() - Remove the specified item (does nothing if the value is not found)
# O(1)
# remove() - Removes the specified element (raises an exception if the value is not found)

# O(1)
# pop() - Removes and returns a random item from the set

# O(S1 + S2)
# union() - Return a set containing the union of sets
# O(S2)
# update() - Update the set with the union of this set and others

# O(S1)
# difference() - Returns a set containing the difference between two sets
# O(S1)
# difference_update() - Removes the items in this set that are also included in another, specified set

# O(min(S1, S2))
# intersection() - Returns a set, that is the intersection of two other sets
# O(S1)
# intersection_update() - Removes the items in this set that are not present in other, specified set

# O(S1 + S2)
# symmetric_difference() - Returns a set with the symmetric differences of two sets
# O(S2)
# symmetric_difference_update() - inserts the symmetric differences from this set and another

# O(min(S1, S2))
# isdisjoint() - Returns whether two sets have a intersection or not
# O(S1)
# issubset() - Returns whether another set contains this set or not
# O(S2)
# issuperset() - Returns whether this set contains another set or not

File: JAVASCRIPT/Data Structures/3. Sets.js
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
// https://www.w3schools.com/js/js_object_sets.asp

let mySet = new Set();

// O(1)
// add() - Adds a new element to the Set
mySet.add(1);
mySet.add(2);
mySet.add(3);
mySet.add(4);
mySet.add(5);
console.log(mySet); // Output: Set(5) { 1, 2, 3, 4, 5 }

// O(1)
// delete() - Removes an element from a Set
mySet.delete(1);
console.log(mySet); // Output: Set(4) { 2, 3, 4, 5 }

// O(N)
// clear() - Removes all elements from a Set
mySet.clear();
console.log(mySet); // Output: Set(0)

// O(1)
// has() - Returns whether an element is in a Set
console.log(mySet.has(1)); // false

// O(1)
// size() - Returns the number of elements in a Set
console.log(mySet.size); // 0

// O(N)
// values() - Returns a new Iterator object containing the values for each element in the Set
mySet.add(1);
mySet.add(2);
mySet.add(3);
mySet.add(4);
mySet.add(5);
console.log(mySet.values()); // Output: [Set Iterator] { 1, 2, 3, 4, 5 }

// O(N)
// entries() - Returns a new Iterator object containing the [key, value] pairs for each element in the Set
console.log(mySet.entries()); // Output: [Set Entries] { [ 1, 1 ], [ 2, 2 ], [ 3, 3 ], [ 4, 4 ], [ 5, 5 ] }

// O(N)
// keys() - Returns a new Iterator object containing the keys for each element in the Set
console.log(mySet.keys()); // Output: [Set Iterator] { 1, 2, 3, 4, 5 }

// O(N)
// forEach() - Calls a function for each element in the Set
mySet.forEach((value) => console.log(value)); // Output: 1 2 3 4 5
